test/helpers/envtest.go

```diff
@@ -0,0 +1,132 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+func init() {
+	// Chdir to project root so the paths to CRDs are stable
+	_, filename, _, _ := goruntime.Caller(0) //nolint
+	dir := path.Join(path.Dir(filename), "..")
+	err := os.Chdir(dir)
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+type TestEnvironment struct {
+	*envtest.Environment
+	manager.Manager
+	client.Client
+
+	Config  *rest.Config
+	scheme  *runtime.Scheme
+	doneMgr chan struct{}
+}
+
+func newTestEnvironment() *TestEnvironment {
+	scheme := scheme.Scheme
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+	utilruntime.Must(bootstrapv1.AddToScheme(scheme))
+
+	return &TestEnvironment{
+		Environment: &envtest.Environment{
+			ErrorIfCRDPathMissing: true,
+			CRDDirectoryPaths: []string{
+				filepath.Join("..", "config", "crd", "bases"),
+				filepath.Join("..", "controlplane", "kubeadm", "config", "crd", "bases"),
+				filepath.Join("..", "bootstrap", "kubeadm", "config", "crd", "bases"),
+			},
+			CRDs: []runtime.Object{
+				external.TestGenericBootstrapCRD,
+				external.TestGenericBootstrapTemplateCRD,
+				external.TestGenericInfrastructureCRD,
+				external.TestGenericInfrastructureTemplateCRD,
+			},
+		},
+		scheme:  scheme,
+		doneMgr: make(chan struct{}),
+	}
+}
+
+func NewTestEnvironment() (*TestEnvironment, error) {
+	env := newTestEnvironment()
+
+	cfg, err := env.Environment.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	mgr, err := manager.New(cfg, manager.Options{
+		Scheme:             env.scheme,
+		MetricsBindAddress: "0",
+		NewCache: func(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+			syncPeriod := 1 * time.Second
+			opts.Resync = &syncPeriod
+			return cache.New(config, opts)
+		},
+	})
+	if err != nil {
+		if stopErr := env.Stop(); err != nil {
+			return nil, kerrors.NewAggregate([]error{err, stopErr})
+		}
+		return nil, err
+	}
+
+	env.Config = cfg
+	env.Manager = mgr
+	env.Client = mgr.GetClient()
+
+	return env, nil
+}
+
+func (t *TestEnvironment) StartManager() error {
+	return t.Manager.Start(t.doneMgr)
+}
+
+func (t *TestEnvironment) Stop() error {
+	close(t.doneMgr)
+	return t.Environment.Stop()
+}
+
+func (t *TestEnvironment) CreateKubeconfigSecret(cluster *clusterv1.Cluster) error {
+	return t.Create(context.TODO(), kubeconfig.GenerateSecret(cluster, kubeconfig.FromEnvTestConfig(t.Config, cluster)))
```

Comment:
To avoid the breaking change, this function could call to the util/kubeconfig one for now